packages/cache/src/internal/tar.ts

```diff
@@ -117,31 +132,43 @@ export async function createTar(
   // --long=#: Enables long distance matching with # bits. Maximum is 30 (1GB) on 32-bit OS and 31 (2GB) on 64-bit.
   // Using 30 here because we also support 32-bit self-hosted runners.
   // Long range mode is added to zstd in v1.3.2 release, so we will not use --long in older version of zstd.
-  function getCompressionProgram(): string[] {
+  async function getCompressionProgram(): Promise<string[]> {
+    const tarPath = await getTarPath([])
+    const BSD_TAR_ZSTD = IS_WINDOWS && tarPath === SystemTarPathOnWindows
     switch (compressionMethod) {
       case CompressionMethod.Zstd:
+        if (BSD_TAR_ZSTD) {
+          return ['-O', '|', 'zstd -T0 --long=30 -o']
+        }
         return [
           '--use-compress-program',
-          IS_WINDOWS ? 'zstd -T0 --long=30' : 'zstdmt --long=30'
+          IS_WINDOWS ? 'zstd -T0 --long=30' : 'zstdmt --long=30',
+          '-cf'
         ]
       case CompressionMethod.ZstdWithoutLong:
-        return ['--use-compress-program', IS_WINDOWS ? 'zstd -T0' : 'zstdmt']
+        if (BSD_TAR_ZSTD) {
+          return ['-O', '|', 'zstd -T0 -o']
+        }
+        return [
+          '--use-compress-program',
+          IS_WINDOWS ? 'zstd -T0' : 'zstdmt',
+          '-cf'
+        ]
       default:
-        return ['-z']
+        return ['-z', '-cf']
     }
   }
   const args = [
     '--posix',
-    ...getCompressionProgram(),
-    '-cf',
-    cacheFileName.replace(new RegExp(`\\${path.sep}`, 'g'), '/'),
     '--exclude',
     cacheFileName.replace(new RegExp(`\\${path.sep}`, 'g'), '/'),
     '-P',
     '-C',
     workingDirectory.replace(new RegExp(`\\${path.sep}`, 'g'), '/'),
     '--files-from',
-    manifestFilename
+    manifestFilename,
+    ...(await getCompressionProgram()),
+    cacheFileName.replace(new RegExp(`\\${path.sep}`, 'g'), '/')
```

Comment:
Does moving this args order have any negative effect on gnutar case or bsdtar on macos case? Essentially it would be good to idea to test all these existing scenarios for regression.

Reply:
Generally it shouldn't affect. In windows case due to pipe symbol it is needed. I would take a look at it